### PR TITLE
[Merged by Bors] - feat: port RepresentationTheory.Basic

### PR DESCRIPTION
**Pull request by mattrobball**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RepresentationTheory/Basic.lean [hunk 1]
```diff
@@ -289,26 +297,43 @@ theorem ofMulAction_apply {H : Type _} [MulAction G H] (g : G) (f : H →₀ k)
...
 #align representation.of_mul_action_apply Representation.ofMulAction_apply
 
-theorem ofMulAction_self_smul_eq_mul (x : MonoidAlgebra k G) (y : (ofMulAction k G G).AsModule) :
-    x • y = (x * y : MonoidAlgebra k G) :=
-  x.inductionOn (fun g => by show as_algebra_hom _ _ _ = _ <;> ext <;> simp)
+-- Porting note: did not need this in ML3; noncomputable because IR check complains
+noncomputable instance :
+    HMul (MonoidAlgebra k G) ((ofMulAction k G G).asModule) (MonoidAlgebra k G) := by
+  change HMul (MonoidAlgebra k G) (MonoidAlgebra k G) (MonoidAlgebra k G)
+  infer_instance
```
Comment (jcommelin) [inline Mathlib/RepresentationTheory/Basic.lean:311]: Please use `inferInstanceAs`

File: Mathlib/RepresentationTheory/Basic.lean [hunk 2]
```diff
@@ -90,73 +91,80 @@ theorem asAlgebraHom_def : asAlgebraHom ρ = (lift k G _) ρ :=
...
-noncomputable instance asModuleModule : Module (MonoidAlgebra k G) ρ.AsModule :=
+noncomputable instance asModuleModule : Module (MonoidAlgebra k G) ρ.asModule :=
   Module.compHom V (asAlgebraHom ρ).toRingHom
 #align representation.as_module_module Representation.asModuleModule
 
-/-- The additive equivalence from the `module (monoid_algebra k G)` to the original vector space
+-- Porting note: ρ.asModule doesn't unfold now
+instance : Module k ρ.asModule := by
+  change Module k V
+  infer_instance
```
[the remark this answers is not in the document]
Comment (mattrobball) [reply Mathlib/RepresentationTheory/Basic.lean]: Done!

File: Mathlib/RepresentationTheory/Basic.lean (hunk 1 above)
Comment (mattrobball) [reply Mathlib/RepresentationTheory/Basic.lean]: Done!

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**